internal/initializer/tls.go

```diff
@@ -0,0 +1,332 @@
+limitations under the License.
+*/
+
+package initializer
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+const (
+	errGenerateCA              = "cannot generate CA certificate"
+	errParseCACertificate      = "cannot parse CA certificate"
+	errParseCAKey              = "cannot parse CA key"
+	errLoadOrGenerateSigner    = "cannot load or generate certificate signer"
+	errDecodeKey               = "cannot decode key"
+	errDecodeCert              = "cannot decode cert"
+	errFmtGetTLSSecret         = "cannot get TLS secret: %s"
+	errFmtCannotCreateOrUpdate = "cannot create or update secret: %s"
+)
+
+const (
+	// RootCACertSecretName is the name of the secret that will store CA certificates and rest of the
+	// certificates created per entities will be signed by this CA
+	RootCACertSecretName = "root-ca-certs"
+)
+
+// TLSCertificateGenerator is an initializer step that will find the given secret
+// and fill its tls.crt, tls.key and ca.crt fields to be used for External Secret
+// Store plugins
+type TLSCertificateGenerator struct {
+	namespace           string
+	caSecretName        string
+	tlsServerSecretName string
+	tlsClientSecretName string
+	subject             string
+	owner               []metav1.OwnerReference
+	certificate         CertificateGenerator
+	log                 logging.Logger
+}
+
+// TLSCertificateGeneratorOption is used to configure TLSCertificateGenerator behavior.
+type TLSCertificateGeneratorOption func(*TLSCertificateGenerator)
+
+// TLSCertificateGeneratorWithLogger returns an TLSCertificateGeneratorOption that configures logger
+func TLSCertificateGeneratorWithLogger(log logging.Logger) TLSCertificateGeneratorOption {
+	return func(g *TLSCertificateGenerator) {
+		g.log = log
+	}
+}
+
+// NewTLSCertificateGenerator returns a new TLSCertificateGenerator.
+func NewTLSCertificateGenerator(ns, caSecret, tlsServerSecret, tlsClientSecret, subject string, owner []metav1.OwnerReference, opts ...TLSCertificateGeneratorOption) *TLSCertificateGenerator {
+	e := &TLSCertificateGenerator{
+		namespace:           ns,
+		caSecretName:        caSecret,
+		tlsServerSecretName: tlsServerSecret,
+		tlsClientSecretName: tlsClientSecret,
+		subject:             subject,
+		owner:               owner,
+		certificate:         NewCertGenerator(),
+		log:                 logging.NewNopLogger(),
+	}
+
+	for _, f := range opts {
+		f(e)
+	}
+	return e
+}
+
+func (e *TLSCertificateGenerator) loadOrGenerateCA(ctx context.Context, kube client.Client, nn types.NamespacedName) (*CertificateSigner, error) {
```

Comment:
I see that several methods of `TLSCertificateGenerator` take a `client.Client` as an argument. Would it make sense for `TLSCertificateGenerator` to have a `client.Client` struct field instead?

Reply:
The reason I kept it as an argument is, the Run function is implementing the Step [interface](https://github.com/crossplane/crossplane/blob/a888ea7657f12d2dd9effa5ef8da8b17b0bea404/internal/initializer/initializer.go#L36).